makerdao/governance-portal-v2
# feat: Poll parameters

### Link to Shortcut ticket:

### What does this PR do?

This PR focuses on the introduction of the governance poll parameters. 
It focuses on the following different areas:
- Change on poll metadata, examples found on the DUX community repo https://github.com/makerdao-dux/community/tree/dux/goerli-polls/governance/polls/poll-parameters
- Change on the UI of poll creation to validate those parameters and inform the user if those are wrong. Add new UI to verify poll markdown before creation
- Unit tests to verify correctness of poll parameters
- Documentation of the new poll parameters
- New poll parameters data model that replaces the previous "Pluralitty / IRV" vote type on the gov portal 👁️ 

### New poll data model

The new poll metadata object is defined with the following properties:

```
poll: {
  ...
  parameters: {
    inputFormat : 'rank-free' / 'single-choice',
    victoryConditions?: ["majority" , "condition-o1 : o1>=10,000"],
    resultDisplay: 'single-vote-breakdown' | 'instant-runoff-breakdown' 
  }
}
```
Old "vote-types" 

This "parameters" property defines how a poll vote has to be rendered in the ui:
- inputFormat determines how the options are presented to the user in the voting process
- victoryConditions: determines what algorithm has to be used to calculate the winner, and additionaly what conditions can be satisfied (some algorithms need extra options)
- resultDisplay: how the result of the votes is represented

## Input format

single-choice | Current Single Choice input UI as used for plurality
-- | --
rank-free | Current Ranked Choice input UI as used for Ranked Choice
choose-free | UI to Select n options from list of options


## Victory conditions:
Victory conditions is an array of values that determine the wining condition of the algorithm

Currently we supported plurality and instant-runoff(IRV)


victory_conditions | summary
-- | --
majority | Condition returns true for the option with more than 50% of the total vote-weight - false for others
plurality | Condition return true for the option with more vote weight than any other option - false for others
instant-runoff | Condition returns true for the option winning when applying the IRV algorthim - false for others
"condition-<option(s)> : <option/value><comparator><option/value>" | Condition returns true for <option(s)> if expression <option/value><comparator><option/value> evaluates to true - false for others
no-victor | Condition returns false for every option (marking a poll that is not supposed to result in a winning option)
default-<option> | Condition returns true for <option> if no other option returns true on all other victory conditions
approval | Condition return true for the option with the most approval - false for others

## Result display:

result_display |  
-- | --
single-vote-breakdown | Displays current breakdown UI for Plurality
instant-runoff-breakdown | Displays current breakdown UI for Ranked Choice
condition-summary | Displays a table that lists each victory condition and shows whether it evaluates to true or false for each option
score-<option/value><operator><option/value> | Displays an output score where score = <option/value><operator><option/value>
approval-breakdown | Displays current breakdown UI for Plurality - but with approval scores.
weight-breakdown | Displays current breakdown UI for Plurality - but with weighted scores.

## Example of the new poll parameters on a poll
We have created several test polls that can be found at:  https://github.com/makerdao-dux/community/tree/dux/goerli-polls/governance/polls/poll-parameters 

In the test polls we have tried to cover all the different scenarios supported right now:
Current IRV: 
```
---

parameters:
  input_format: rank-free
  victory_conditions:
      - { type : instant-runoff }
  result_display: instant-runoff-breakdown
version: v0.0.1    

---

```
Current plurality:
```
---

parameters:
    input_format: single-choice
    victory_conditions:
        - { type : plurality }
    result_display: single-vote-breakdown
version: v0.0.1    

---

```

Not supported:
Majority with conditions:
```
---

parameters:
  input_format: single-choice
  victory_conditions:
    - { type : majority: options: [1,2,3,4] }
    - { type : comparison, options: [0, 1, 4], comparator : '>=10000' }
    - { type : default, value : 2 }
  result_display: single-vote-breakdown
version: v0.0.1    

---

```


### Steps for testing:
- Go to goerli
- check all the latest polls: TBD links
- check vote input , calculation and vote display difference

### Screenshots (if relevant):
- Polling create
![image](https://user-images.githubusercontent.com/1152768/172863666-0fb813fc-898a-4a29-8c29-2386c8da29f5.png)


### Any additional helpful information?:
Notion notes https://www.notion.so/makerdux/Gov-portal-tags-66385b6696c046feb8045a60a255a541
Poll parameters definitions : https://docs.google.com/spreadsheets/d/1TPtA3f8FEfADThzueGjA8Agbs6fIeD1vEDhrGbtskmE/edit#gid=640577726
